### PR TITLE
Coins pallet

### PR DESCRIPTION
**Pull request by akildemir**
Replaces `pallet_tokens`, `pallet_balances` and `tokens_pallet` with a custom single `coins_pallet` .

### CODE REVIEW

File: substrate/client/src/serai/coins.rs [hunk 1]
```diff
@@ -22,32 +20,29 @@ impl<'a> SeraiCoins<'a> {
...
-    > = self.0.storage("System", "Account", Some(vec![scale_value(address)])).await?;
-    Ok(data.map(|data| data.data.free).unwrap_or(0))
+    let amount: Option<SubstrateAmount> = self
+      .0
+      .storage(PALLET, "Coins", Some(vec![scale_value(address), scale_value(Coin::Serai)]))
+      .await?;
+    Ok(amount.unwrap_or(0))
   }
 
+  // TODO: shouldn't these token functions be renamed to coins now?
```
Comment (kayabaNerve) [inline substrate/client/src/serai/coins.rs]: Yep. We should remove `sri_` specific methods and solely have coin_*method* methods (which can be passed Coin::Serai as an argument).

File: substrate/coins/pallet/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,47 @@
+[package]
+name = "serai-coins-pallet"
+version = "0.1.0"
+description = "Coins pallet for Serai"
+license = "AGPL-3.0-only"
+repository = "https://github.com/serai-dex/serai/tree/develop/substrate/staking/pallet"
+authors = ["Luke Parker <lukeparker5132@gmail.com>"]
```
Comment (kayabaNerve) [inline substrate/coins/pallet/Cargo.toml]: You should be the author, actually :p

File: substrate/coins/pallet/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+  pub enum Error<T> {
+    AmountOverflowed,
+    NotEnoughFunds,
+  }
+
+  #[pallet::event]
+  #[pallet::generate_deposit(fn deposit_event)]
+  pub enum Event<T: Config> {
+    Mint { address: SeraiAddress, balance: Balance },
+    Burn { address: SeraiAddress, balance: Balance, instruction: OutInstruction },
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: Do we have an `OutInstructionWithBalance`? We do for `InInstruction`.

File: substrate/coins/pallet/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+    SriBurn { address: SeraiAddress, amount: Amount },
+    Transfer { from: SeraiAddress, to: SeraiAddress, balance: Balance },
+  }
+
+  #[pallet::pallet]
+  pub struct Pallet<T>(PhantomData<T>);
+
+  /// The amount of funds each account has.
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: "amount of coins"

File: substrate/coins/pallet/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+  #[pallet::storage]
+  #[pallet::getter(fn coins)]
+  pub type Coins<T: Config> = StorageDoubleMap<
+    _,
+    Blake2_128Concat,
+    T::AccountId,
+    Identity,
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs:67]: We should have a comment justifying the use of Identity due to Coin being a non-manipulatable fixed-space ID.

File: substrate/coins/pallet/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+  #[pallet::genesis_build]
+  impl<T: Config> BuildGenesisConfig for GenesisConfig<T> {
+    fn build(&self) {
+      // initialize the supply of the coins
+      for c in COINS.iter() {
+        Supply::<T>::set(c, 0);
+      }
+
+      // initialize the genesis accounts
+      for (account, coin, amount) in self.accounts.iter() {
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: Use the Balance type?

File: substrate/coins/pallet/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+        Coins::<T>::set(account, coin, Some(*amount));
+        let current_supply = Supply::<T>::get(coin);
+        Supply::<T>::set(coin, current_supply.checked_add(*amount).unwrap());
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: These lines should be replaced with `mint`. They won't emit events right now *and* overwrite balance due to non-usage of `mint`.

File: substrate/coins/pallet/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+    Transfer { from: SeraiAddress, to: SeraiAddress, balance: Balance },
+  }
+
+  #[pallet::pallet]
+  pub struct Pallet<T>(PhantomData<T>);
+
+  /// The amount of funds each account has.
+  #[pallet::storage]
+  #[pallet::getter(fn coins)]
+  pub type Coins<T: Config> = StorageDoubleMap<
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: Would `Balances` be a better name for this? It's not a StorageValue of all active Coins.

File: substrate/coins/pallet/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+  impl<T: Config> Pallet<T> {
+    /// Returns the balance of a given account for `coin`.
+    pub fn balance(of: &T::AccountId, coin: Coin) -> SubstrateAmount {
+      Self::coins(of, coin).unwrap_or(0)
+    }
+
+    fn increase_supply(coin: Coin, by: SubstrateAmount) -> Result<(), Error<T>> {
+      // update the supply
+      let new_supply = Self::supply(coin).checked_add(by).ok_or(Error::<T>::AmountOverflowed)?;
+      Supply::<T>::set(coin, new_supply);
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: Shouldn't we inline this into mint? I'm unsure why we'd have another caller.

File: substrate/coins/pallet/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+      Ok(())
+    }
+
+    fn decrease_supply(coin: Coin, by: SubstrateAmount) -> Result<(), Error<T>> {
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: Saqme comment as increase_supply.

File: substrate/coins/pallet/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+      // update the supply
+      let new_supply = Self::supply(coin).checked_sub(by).ok_or(Error::<T>::AmountOverflowed)?;
+      Supply::<T>::set(coin, new_supply);
+      Ok(())
+    }
+
+    /// Mints amount at the given account, errors if amount overflows.
+    pub fn mint(at: &T::AccountId, coin: Coin, amount: SubstrateAmount) -> Result<(), Error<T>> {
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: at -> to?

File: substrate/coins/pallet/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+      // don't waste time if amount 0.
+      if amount == 0 {
+        return Ok(());
+      }
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: Ironically enough, I think this check will waste more time than not having this check would. This'll be called by emissions, which should be non-zero, and InInstructions. The processor will drop anything that's dust (such as 0 transfers).

You're welcome to keep it as an idea of good practice though.

File: substrate/coins/pallet/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+      Coins::<T>::set(at, coin, Some(new_amount));
+      Self::increase_supply(coin, amount)?;
+
+      let balance = Balance { coin, amount: Amount(amount) };
+      Self::deposit_event(Event::Mint { address: SeraiAddress(at.0), balance });
+      Ok(())
+    }
+
+    /// Burns amount at the given account, errors if not enough funds to burn.
+    pub fn burn_internal(
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: Shouldn't be pub.

File: substrate/coins/pallet/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+  impl<T: Config> Default for GenesisConfig<T> {
+    fn default() -> Self {
+      GenesisConfig { accounts: Default::default() }
+    }
+  }
+
+  #[pallet::error]
+  pub enum Error<T> {
+    AmountOverflowed,
+    NotEnoughFunds,
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: NotEnoughCoins

File: substrate/coins/pallet/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+
+      // sub amount from account
+      let new_amount =
+        Self::coins(at, coin).unwrap_or(0).checked_sub(amount).ok_or(Error::<T>::NotEnoughFunds)?;
+
+      if new_amount == 0 {
+        Coins::<T>::remove(at, coin);
+      } else {
+        Coins::<T>::set(at, coin, Some(new_amount));
+      }
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: decrease_balance_internal to share this with transfer? Same commentary for increase_balance_internal.

File: substrate/coins/pallet/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,314 @@
...
+      // something should have gone horribly wrong if we error on this.
+      if Self::decrease_supply(coin, amount).is_err() {
+        panic!("we tried to burn more assets than what we have");
+      }
+
+      if coin == Coin::Serai {
+        Self::deposit_event(Event::SriBurn { address: SeraiAddress(at.0), amount: Amount(amount) });
+      } else {
+        let balance = Balance { coin, amount: Amount(amount) };
+        // TODO: error for no instruction?
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: I'd move this `if` statement into the caller so `burn_internal` is InInstruction-agnostic and we don't have this potential error case (even if our static code never makes it reachable). Then we'd offer burn and burn_external, one not taking an instruction, one taking.

File: substrate/coins/pallet/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+    }
+
+    #[pallet::call_index(1)]
+    #[pallet::weight((0, DispatchClass::Normal))] // TODO
+    pub fn burn(
+      o: OriginFor<T>,
+      balance: Balance,
+      instruction: Option<OutInstruction>,
+    ) -> DispatchResult {
+      let at = ensure_signed(o)?;
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: I'd still call this from.

File: substrate/coins/pallet/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,314 @@
...
+      Ok(())
+    }
+  }
+
+  #[pallet::call]
+  impl<T: Config> Pallet<T> {
+    #[pallet::call_index(0)]
+    #[pallet::weight((0, DispatchClass::Normal))] // TODO
+    pub fn transfer(
+      o: OriginFor<T>,
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: This `o` gets compiled into an enum member (`Pallet::Call::transfer { o: OriginFor<T> }`). It'd be polite to call it `origin`-proper.

File: substrate/coins/pallet/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,314 @@
...
+      instruction: Option<OutInstruction>,
+    ) -> DispatchResult {
+      let at = ensure_signed(o)?;
+      Self::burn_internal(&at, balance.coin, balance.amount.0, instruction)?;
+      Ok(())
+    }
+  }
+
+  #[derive(Default)]
+  pub struct Imbalance {
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: Is there a reason not to use SubstrateAmount/Option<SubstrateAmount> directly?

File: substrate/coins/pallet/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,314 @@
...
+  }
+
+  impl<T: Config> OnChargeTransaction<T> for Pallet<T> {
+    type Balance = SubstrateAmount;
+    type LiquidityInfo = Option<Imbalance>;
+
+    /// Before the transaction is executed the payment of the transaction fees
+    /// need to be secured.
+    ///
+    /// Note: The `fee` already includes the `tip`.
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: These comments, directly copied from the trait docs, should not be here.

File: substrate/coins/pallet/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,314 @@
...
+    fn withdraw_fee(
+      who: &T::AccountId,
+      _call: &T::RuntimeCall,
+      _dispatch_info: &DispatchInfoOf<T::RuntimeCall>,
+      fee: Self::Balance,
+      _tip: Self::Balance,
+    ) -> Result<Self::LiquidityInfo, TransactionValidityError> {
+      // TODO: Could we have a 0 fee tx? Error in this case?
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: Yeah, I'd just return Ok as you do.

File: substrate/coins/pallet/src/lib.rs [hunk 22]
```diff
@@ -0,0 +1,314 @@
...
+        Err(_) => Err(InvalidTransaction::Payment.into()),
+        Ok(()) => Ok(Some(Imbalance { amount: fee })),
+      }
+    }
+
+    /// After the transaction was executed the actual fee can be calculated.
+    /// This function should refund any overpaid fees and optionally deposit
+    /// the corrected amount.
+    ///
+    /// Note: The `fee` already includes the `tip`.
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: Same comment as I gave to the prior comment.

File: substrate/coins/primitives/src/lib.rs [hunk 23]
```diff
@@ -12,7 +12,7 @@ use scale_info::TypeInfo;
 
 use serai_primitives::{Balance, SeraiAddress, ExternalAddress, Data, pallet_address};
 
-pub const ADDRESS: SeraiAddress = pallet_address(b"Tokens");
+pub const FEE_ACCOUNT: SeraiAddress = pallet_address(b"FeeAccount");
```
Comment (kayabaNerve) [inline substrate/coins/primitives/src/lib.rs]: Should we rename pallet_address to system_address?

File: substrate/in-instructions/pallet/src/lib.rs [hunk 24]
```diff
@@ -73,10 +73,14 @@ pub mod pallet {
   impl<T: Config> Pallet<T> {
     fn execute(instruction: InInstructionWithBalance) -> Result<(), ()> {
       match instruction.instruction {
-        InInstruction::Transfer(address) => Tokens::<T>::mint(address, instruction.balance),
+        InInstruction::Transfer(address) => {
+          // TODO: pass balance directly here instead of coin and amount?
+          // but coins pallet deals only with SubstrateAmount and Balance type has Amount type.
+          Coins::<T>::mint(&address.into(), instruction.balance.coin, instruction.balance.amount.0)
```
Comment (kayabaNerve) [inline substrate/in-instructions/pallet/src/lib.rs]: Pass Balance, yeah

File: substrate/node/src/chain_spec.rs [hunk 25]
```diff
@@ -34,24 +34,10 @@ fn testnet_genesis(
...
-        .iter()
-        .map(|coin| (*coin, TOKENS_ADDRESS.into(), true, 1))
-        .collect(),
-      metadata: vec![
-        (Coin::Bitcoin, b"Bitcoin".to_vec(), b"BTC".to_vec(), 8),
-        // Reduce to 8 decimals to feasibly fit within u64 (instead of its native u256)
-        (Coin::Ether, b"Ether".to_vec(), b"ETH".to_vec(), 8),
-        (Coin::Dai, b"Dai Stablecoin".to_vec(), b"DAI".to_vec(), 8),
-        (Coin::Monero, b"Monero".to_vec(), b"XMR".to_vec(), 12),
-      ],
```
Comment (kayabaNerve) [inline substrate/node/src/chain_spec.rs:53]: Should we move these to a few new functions under the Coin enum?

File: substrate/runtime/src/lib.rs [hunk 26]
```diff
@@ -217,7 +201,7 @@ impl system::Config for Runtime {
   type OnKilledAccount = ();
   type OnSetCode = ();
 
-  type AccountData = balances::AccountData<SubstrateAmount>;
+  type AccountData = SubstrateAmount;
```
Comment (kayabaNerve) [inline substrate/runtime/src/lib.rs]: This should be `()` now, I believe.

File: substrate/coins/pallet/src/lib.rs [hunk 27]
```diff
@@ -0,0 +1,314 @@
...
+
+      if coin == Coin::Serai {
+        Self::deposit_event(Event::SriBurn { address: SeraiAddress(at.0), amount: Amount(amount) });
+      } else {
+        let balance = Balance { coin, amount: Amount(amount) };
+        // TODO: error for no instruction?
+        Self::deposit_event(Event::Burn {
+          address: SeraiAddress(at.0),
+          balance,
+          instruction: instruction.unwrap(),
```
Comment (kayabaNerve) [inline substrate/coins/pallet/src/lib.rs]: This isn't valid. burn is user-facing. A user can call burn, calling burn_internal, with a non-Serai coin and None for the instruction, causing a panic.

**Review by kayabaNerve — COMMENTED** [review]
Largely lints :+1:

File: substrate/coins/pallet/src/lib.rs (hunk 19 above)
Comment (akildemir) [reply substrate/coins/pallet/src/lib.rs]: yes. The only requirement for `OnChargeTransaction::LiquidityInfo` type is that should be something that implements `Default` trait and can't derive `Default` on plain u64. Hence why it is wrapped into a struct.

File: substrate/coins/primitives/src/lib.rs (hunk 23 above)
Comment (akildemir) [reply substrate/coins/primitives/src/lib.rs]: makes sense

File: substrate/coins/pallet/src/lib.rs (hunk 19 above)
Comment (kayabaNerve) [reply substrate/coins/pallet/src/lib.rs]: Ah, then Option<u64> should work.

**Comment by kayabaNerve** [comment]
CI failure is unrelated and fixed in https://github.com/serai-dex/serai/commit/041ed4617112fd350e09744b380b8e89fb5c62e8

**Pull request merged**